feat: improved manual (#9940)
emberjs/data: guides/manual/1-overview.md

<table>
  <tr>
    <td align="center" width="300"></td>
    <td align="center" width="300">
   
[❖ Table of Contents](./0-index.md)</td>
   <td align="center" width="300">

[Making Requests →](./3-requests.md)</td>
  </tr>
</table>

## Overview

WarpDrive is the data framework for building ambitious applications.

What do we mean by ambitious? WarpDrive is ideal for applications looking to be best-in-class: whether that's a small todo app, e-commerce, a social app, or an enterprise b2b software solution.

> The capabilities WarpDrive provides are designed to seamlessly handle and simplify the hardest parts of state management when building an app, helping you focus on creating the features and user experiences that drive value.

Author: added (directly) since I was tweaking other stuff

> Its innovative approach to [fine grained reactivity](https://dev.to/ryansolid/a-hands-on-introduction-to-fine-grained-reactivity-3ndf) enables rapidly developing robust, performant web applications using any [Signals](https://github.com/tc39/proposal-signals#readme) compatible framework such as [Ember](https://guides.emberjs.com/release/in-depth-topics/autotracking-in-depth/), [Svelte](https://svelte.dev/docs/svelte/what-are-runes), [Angular](https://angular.dev/guide/signals), [Vue.js](https://vuejs.org/guide/extras/reactivity-in-depth.html), [SolidJS](https://www.solidjs.com/tutorial/introduction_signals),

Reviewer: ```suggestion
WarpDrive's innovative approach to [fine grained reactivity](https://dev.to/ryansolid/a-hands-on-introduction-to-fine-grained-reactivity-3ndf) enables rapidly developing robust, performant web applications using any [Signals](https://github.com/tc39/proposal-signals#readme) compatible framework such as [Ember](https://guides.emberjs.com/release/in-depth-topics/autotracking-in-depth/), [Svelte](https://svelte.dev/docs/svelte/what-are-runes), [Angular](https://angular.dev/guide/signals), [Vue.js](https://vuejs.org/guide/extras/reactivity-in-depth.html), [SolidJS](https://www.solidjs.com/tutorial/introduction_signals),
```

Author: Used "Our" instead of "Its", felt it was more direct/conversational/approachable

[Preact](https://preactjs.com/guide/v10/signals/) or [Lit](https://lit.dev/docs/data/signals/).

```hbs
  Hello {{@user.name}}!
```

By building around the same interface as the [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API), WarpDrive makes powerful request management features like caching, deduping and data normalization feel simple to use. We pair this with [reactive control flow](./concepts/reactive-control-flow.md) to give apps the ability to declaratively derive states with safety.

> ```glimmer-ts

Reviewer: Loving the template-only example. Really helps to illustrate how little you have to do to integrate.

import { Request } from '@warp-drive/ember';
import { findRecord } from '@ember-data/json-api/request';
import { Spinner } from './spinner';

export default <template>
  <Request @query={{findRecord "user" "1"}}>
    <:error as |error state|>
      <h2>Whoops!</h2>
      <h3>{{error.message}}</h3>
      <button {{on "click" state.retry}}>Try Again?</button>
    </:error>

    <:content as |result|>
      Hello {{result.data.name}}!
    </:content>

    <:loading as |state|>
      <Spinner @progress={{state.completedRatio}} />
      <button {{on "click" state.abort}}>Cancel?</button>
    </:loading>
  </Request>
</template>
```

> Web clients are like high-latency, remotely distributed, often-stale partial replicas of server state - but requests using WarpDrive utilize an [advanced relational cache](./5-caching.md) that simplifies these problems, solving them when it can and providing intelligent escape valves for when it can't so that no matter what you can quickly get the data you need in the right state.

Reviewer: This is a pretty important, but very dense paragraph. I wonder if we should illustrate it more clearly for less experienced devs. (Will continue this thought after my computer restarts to install upgrades...)

> Web clients are like high-latency, remotely distributed, often-stale partial replicas of server state - but requests using WarpDrive utilize an [advanced relational cache](./5-caching.md) that simplifies these problems, solving them when it can and providing intelligent escape valves for when it can't so that no matter what you can quickly get the data you need in the right state.

Reviewer: ```suggestion
**Web clients are like high-latency, remotely distributed, often-stale partial replicas of server state**. WarpDrive provides an [advanced relational cache](./5-caching.md) that simplifies these problems--solving them when it can and providing intelligent escape valves for when it can't. No matter what, you can quickly **get the data you need in the right state**.
```

Author: I wavered on whether I like all of this change, none of this change, or some of this change. I think in net I like all of it and it's a good change to make the next pass through from. Added change manually.

```ts
import { withDefaults } from '@warp-drive/schema-record';

store.schema.registerResource(
  withDefaults({
    type: 'user',
    fields: [
      { kind: 'field', name: 'firstName' },
      { kind: 'field', name: 'lastName' },
      { kind: 'field', name: 'birthday', type: 'luxon-date' },
      {
        kind: 'derived',
        name: 'age',
        type: 'duration',
        options: { field: 'birthday', format: 'y' }
      },
      {
        kind: 'derived',
        name: 'fullName',
        type: 'concat',
        options: { fields: ['firstName', 'lastName'], separator: ' ' }
      },
    ]
  })
)
```

[Mutation](./concepts/mutations.md) is handled within controlled contexts. The data to edit is "checked out" for editing, giving access to a mutable version. Local edits are seamlessly preserved if the user navigates away and returns without saving, and the changes are buffered from appearing elsewhere in your app until they are also committed to the server.

```ts
import { Checkout } from '@warp-drive/schema-record';

// ...

> const editable = await user[Checkout]();

Reviewer: This might be confusing for junior devs. Can we provide an API like `await checkout(user)`?

Author: I dislike it for a different reason (requiring the import). I've considered adding `await store.checkoutRecord(user)` but checkout is an implementation detail of SchemaRecord currently, we'd have to consider whether this is a permanent addition of a feature that all potential record classes must support and if so what the contract is

editable.firstName = 'Chris';
```

> WarpDrive is only semi-opinionated about your API. Almost every API is compatible just by authoring a [request handler](./concepts/handlers.md) to ensure that the responses are normalized into the cache format.

Reviewer: ```suggestion
WarpDrive is only semi-opinionated about your API. We provide pre-build [request handlers](./concepts/handlers.md) for common API formats, and you can author custom request handlers to ensure that custom API formats are normalized into the cache format.
```

Author: we don't provide pre-built handlers for any format (and its unlikely we ever do)

```ts
const NormalizeKeysHandler = {
  request(context, next) {
    return next(context.request).then((result) => {
      return convertUnderscoredKeysToCamelCase(result.content);
    });
  }
}
```

The [cache](https://github.com/emberjs/data/blob/main/packages/core-types/src/cache.ts) itself is also pluggable if using a different cache format would work better for the demands of your API.

> ```ts
> import Store from '@ember-data/store';
> import { AppCache } from './my-cache';
> 
> class AppStore extends Store {
>   createCache(capabilities) {
>     return new AppCache(capabilities);
>   }
> }
> ```

Reviewer: ```suggestion
```ts
import Store from '@ember-data/store';
import { CustomCache } from './my-cache';

class AppStore extends Store {
  createCache(capabilities) {
    return new CustomCache(capabilities);
  }
}
```
```

Author: added directly since this had a formatting error
